src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/CodeActions/CodeActionEndpoint.cs

```diff
@@ -22,42 +24,82 @@ internal class CodeActionEndpoint : ICodeActionHandler
         private readonly IEnumerable<RazorCodeActionProvider> _providers;
         private readonly ForegroundDispatcher _foregroundDispatcher;
         private readonly DocumentResolver _documentResolver;
-        private readonly ILogger _logger;
+        private readonly ILanguageServer _languageServer;
 
         private CodeActionCapability _capability;
 
+        private bool? _supportsCodeActionResolve = null;
+
         public CodeActionEndpoint(
             IEnumerable<RazorCodeActionProvider> providers,
             ForegroundDispatcher foregroundDispatcher,
             DocumentResolver documentResolver,
-            ILoggerFactory loggerFactory)
+            ILanguageServer languageServer)
         {
-            if (loggerFactory is null)
-            {
-                throw new ArgumentNullException(nameof(loggerFactory));
-            }
-
             _providers = providers ?? throw new ArgumentNullException(nameof(providers));
             _foregroundDispatcher = foregroundDispatcher ?? throw new ArgumentNullException(nameof(foregroundDispatcher));
             _documentResolver = documentResolver ?? throw new ArgumentNullException(nameof(documentResolver));
-            _logger = loggerFactory.CreateLogger<CodeActionEndpoint>();
+            _languageServer = languageServer ?? throw new ArgumentNullException(nameof(languageServer));
+        }
+
+        internal bool SupportsCodeActionResolve
+        {
+            private get
+            {
+                if (_supportsCodeActionResolve is null)
+                {
+                    var languageServerInstance = _languageServer as LanguageServerInstance;
+                    var extendableClientCapabilities = languageServerInstance?.ClientSettings?.Capabilities as ExtendableClientCapabilities;
+                    _supportsCodeActionResolve = extendableClientCapabilities?.SupportsCodeActionResolve ?? false;
```

Comment:
Do we expect this value to change on the fly or can this just be precomputed once in the constructor?

Reply:
The language server isn't yet initialized in the constructor. Must be done later on, but only once (assume it doesn't change on the fly).